### DISCUSSION TITLE
Two tokens with prepareHeaders

### DISCUSSION THREAD

**Discussion by User6531** [reduxjs/redux-toolkit#3718] [Q&A] [answered]
Hello everyone, I have a question about the `prepareHeaders`. According to the [documentation](https://redux-toolkit.js.org/rtk-query/api/fetchBaseQuery#prepareheaders), it seems straightforward: I just need to retrieve the token from the storage and then add it to the headers. However, in my situation, I have two different tokens for different types of requests. So, I'd like to share my solution with you and get your opinion on it. Thank you!

```tsx
const baseQueryInitial = fetchBaseQuery({
  baseUrl: mainBaseUrl,
  prepareHeaders: async (headers, api) => {
    const { getState, endpoint } = api;
    const state = getState() as RootState;

    if (endpointsWithB2CToken.includes(endpoint)) {
      const appConfig = selectAppConfig(state);
      const b2cToken: string = isEmptyObject(appConfig)
        ? ""
        : await acquireAccessToken(appConfig);

      headers.set("authorization", `Bearer ${b2cToken}`);
    } else {
      const internalToken = selectInternalToken(state);

      headers.set("authorization", `Bearer ${internalToken}`);
    }

    return headers;
  },
});
```

**Comment by EskiMojo14** [comment, marked as answer]
seems fine to me - i imagine this is the main reason endpoint is passed as part of the api object.

**Reply by User6531** [reply]
😅 Oh, okay. Thank you for your feedback.

**Comment by User6531** [comment]
What do you think about using this approach? I will drill the property through headers, so I will put the type of token that should be used in the query, and then in the `prepareHeaders` function, I will read it and delete it from the headers.


```tsx
builder.query({
  query: (customerId) => ({
    url: '...url'
    headers: {
      tokenType: "first",
    },
  }),
});


prepareHeaders: async (headers) => {
  const tokenType = headers.get("tokenType");
  headers.delete("tokenType");

  if (tokenType === 'first') {
    headers.set("authorization", `Bearer ${firstToken}`);
  } else {
    headers.set("authorization", `Bearer ${secondToken}`);
  }

  return headers;
}

```

**Reply by EskiMojo14** [reply]
sure, that works too 🙂
